Packer crash when registering AMI: unexpected EOF / panic makeslice: cap out of range

packer 1.0.2 (run from the official docker image),

building using amazon-ebssurrogate,
the build itself seems to succeed but it errors when attempting to register AMI:

```
==> amazon-ebssurrogate: Registering the AMI...
2017/06/23 10:13:01 ui: ==> amazon-ebssurrogate: Registering the AMI...
2017/06/23 10:13:01 ui: ==> amazon-ebssurrogate: Terminating the source AWS instance...
==> amazon-ebssurrogate: Terminating the source AWS instance...
2017/06/23 10:13:02 packer: 2017/06/23 10:13:02 Waiting for state to become: terminated
2017/06/23 10:13:02 packer: 2017/06/23 10:13:02 Using 2s as polling delay (change with AWS_POLL_DELAY_SECONDS)
2017/06/23 10:13:02 packer: 2017/06/23 10:13:02 Allowing 300s to complete (change with AWS_TIMEOUT_SECONDS)
2017/06/23 10:13:08 ui: ==> amazon-ebssurrogate: Deleting temporary security group...
==> amazon-ebssurrogate: Deleting temporary security group...
2017/06/23 10:13:08 ui: ==> amazon-ebssurrogate: Deleting temporary keypair...
==> amazon-ebssurrogate: Deleting temporary keypair...
2017/06/23 10:13:08 packer: Build 'amazon-ebssurrogate' errored: unexpected EOF

==> Some builds didn't complete successfully and had errors:
--> amazon-ebssurrogate: unexpected EOF

==> Builds finished but no artifacts were created.
panic: runtime error: makeslice: cap out of range
2017/06/23 10:13:08 packer:
2017/06/23 10:13:08 packer: goroutine 30 [running]:
2017/06/23 10:13:08 packer: github.com/hashicorp/packer/builder/amazon/ebssurrogate.(*StepRegisterAMI).Run(0xc420249d50, 0x22b0740, 0xc4202b48a0, 0x0)
2017/06/23 10:13:08 packer: 	/Users/mwhooker/go/src/github.com/hashicorp/packer/builder/amazon/ebssurrogate/step_register_ami.go:28 +0x1ea
2017/06/23 10:13:08 packer: github.com/hashicorp/packer/vendor/github.com/mitchellh/multistep.(*BasicRunner).Run(0xc4202b0880, 0x22b0740, 0xc4202b48a0)
2017/06/23 10:13:08 packer: 	/Users/mwhooker/go/src/github.com/hashicorp/packer/vendor/github.com/mitchellh/multistep/basic_runner.go:70 +0x1e2
2017/06/23 10:13:08 packer: github.com/hashicorp/packer/builder/amazon/ebssurrogate.(*Builder).Run(0xc42025c000, 0x22b3420, 0xc4202b2560, 0x22ab3c0, 0xc4202ac4d0, 0x22b1e80, 0xc420066260, 0xc4202b40c0, 0x0, 0xc420016d20, ...)
2017/06/23 10:13:08 packer: 	/Users/mwhooker/go/src/github.com/hashicorp/packer/builder/amazon/ebssurrogate/builder.go:236 +0x1611
2017/06/23 10:13:08 packer: github.com/hashicorp/packer/packer/rpc.(*BuilderServer).Run(0xc42000b140, 0x1, 0xc4202ac430, 0x0, 0x0)
2017/06/23 10:13:08 packer: 	/Users/mwhooker/go/src/github.com/hashicorp/packer/packer/rpc/builder.go:93 +0x1ea
2017/06/23 10:13:08 packer: reflect.Value.call(0xc420016d20, 0xc42000c0f0, 0x13, 0x18de246, 0x4, 0xc4202a5f20, 0x3, 0x3, 0xc4202dbce0, 0x0, ...)
2017/06/23 10:13:08 packer: 	/usr/local/Cellar/go/1.8.3/libexec/src/reflect/value.go:434 +0x91f
2017/06/23 10:13:08 packer: reflect.Value.Call(0xc420016d20, 0xc42000c0f0, 0x13, 0xc42031ef20, 0x3, 0x3, 0xc42031ef60, 0x70ecf3, 0x154dd80)
2017/06/23 10:13:08 packer: 	/usr/local/Cellar/go/1.8.3/libexec/src/reflect/value.go:302 +0xa4
2017/06/23 10:13:08 packer: net/rpc.(*service).call(0xc420018800, 0xc4200187c0, 0xc420012980, 0xc420128800, 0xc42000b860, 0x15575c0, 0xc4202ac42c, 0x18a, 0x1523860, 0xc4202ac430, ...)
2017/06/23 10:13:08 packer: 	/usr/local/Cellar/go/1.8.3/libexec/src/net/rpc/server.go:387 +0x144
2017/06/23 10:13:08 packer: created by net/rpc.(*Server).ServeCodec
2017/06/23 10:13:08 packer: 	/usr/local/Cellar/go/1.8.3/libexec/src/net/rpc/server.go:481 +0x404```
```
Full log and template/script can be found in https://gist.github.com/mlitvin/653fdcbf89e3b09c1560f3cacab9fc30

--- a/builder/amazon/ebssurrogate/step_register_ami.go
+++ b/builder/amazon/ebssurrogate/step_register_ami.go
@@ -25,7 +25,8 @@ func (s *StepRegisterAMI) Run(state multistep.StateBag) multistep.StepAction {
 
 	ui.Say("Registering the AMI...")
 
-	blockDevicesExcludingRoot := make([]*ec2.BlockDeviceMapping, 0, len(s.BlockDevices)-1)
+	// Defensive coding to make sure we only add the root volume once
+	blockDevicesExcludingRoot := make([]*ec2.BlockDeviceMapping, 0, len(s.BlockDevices))
 	for _, blockDevice := range s.BlockDevices {
 		if *blockDevice.DeviceName == s.RootDevice.SourceDeviceName {
 			continue

--- a/builder/amazon/ebssurrogate/step_register_ami.go
+++ b/builder/amazon/ebssurrogate/step_register_ami.go
@@ -25,17 +25,7 @@ func (s *StepRegisterAMI) Run(state multistep.StateBag) multistep.StepAction {
 
 	ui.Say("Registering the AMI...")
 
-	// Defensive coding to make sure we only add the root volume once
-	blockDevicesExcludingRoot := make([]*ec2.BlockDeviceMapping, 0, len(s.BlockDevices))
-	for _, blockDevice := range s.BlockDevices {
-		if *blockDevice.DeviceName == s.RootDevice.SourceDeviceName {
-			continue
-		}
-
-		blockDevicesExcludingRoot = append(blockDevicesExcludingRoot, blockDevice)
-	}
-
-	blockDevicesExcludingRoot = append(blockDevicesExcludingRoot, s.RootDevice.createBlockDeviceMapping(snapshotId))
+	blockDevicesExcludingRoot := DeduplicateRootVolume(s.BlockDevices, s.RootDevice, snapshotId)
 
 	registerOpts := &ec2.RegisterImageInput{
 		Name:                &config.AMIName,
@@ -125,4 +115,19 @@ func (s *StepRegisterAMI) Cleanup(state multistep.StateBag) {
 		ui.Error(fmt.Sprintf("Error deregistering AMI, may still be around: %s", err))
 		return
 	}
+}
+
+func DeduplicateRootVolume(BlockDevices []*ec2.BlockDeviceMapping, RootDevice RootBlockDevice, snapshotId string) []*ec2.BlockDeviceMapping {
+	// Defensive coding to make sure we only add the root volume once
+	blockDevicesExcludingRoot := make([]*ec2.BlockDeviceMapping, 0, len(BlockDevices))
+	for _, blockDevice := range BlockDevices {
+		if *blockDevice.DeviceName == RootDevice.SourceDeviceName {
+			continue
+		}
+
+		blockDevicesExcludingRoot = append(blockDevicesExcludingRoot, blockDevice)
+	}
+
+	blockDevicesExcludingRoot = append(blockDevicesExcludingRoot, RootDevice.createBlockDeviceMapping(snapshotId))
+	return blockDevicesExcludingRoot
 }

--- a/builder/amazon/ebssurrogate/step_register_ami_test.go
+++ b/builder/amazon/ebssurrogate/step_register_ami_test.go
@@ -1,0 +1,37 @@
+package ebssurrogate
+
+import (
+	"testing"
+
+	"github.com/aws/aws-sdk-go/service/ec2"
+)
+
+func GetStringPointer() *string {
+	tmp := "/dev/name"
+	return &tmp
+}
+
+func GetTestDevice() *ec2.BlockDeviceMapping {
+	TestDev := ec2.BlockDeviceMapping{
+		DeviceName: GetStringPointer(),
+	}
+	return &TestDev
+}
+
+func TestStepRegisterAmi_DeduplicateRootVolume(t *testing.T) {
+	TestRootDevice := RootBlockDevice{}
+	TestRootDevice.SourceDeviceName = "/dev/name"
+
+	blockDevices := []*ec2.BlockDeviceMapping{}
+	blockDevicesExcludingRoot := DeduplicateRootVolume(blockDevices, TestRootDevice, "12342351")
+	if len(blockDevicesExcludingRoot) != 1 {
+		t.Fatalf("Unexpected length of block devices list")
+	}
+
+	TestBlockDevice := GetTestDevice()
+	blockDevices = append(blockDevices, TestBlockDevice)
+	blockDevicesExcludingRoot = DeduplicateRootVolume(blockDevices, TestRootDevice, "12342351")
+	if len(blockDevicesExcludingRoot) != 1 {
+		t.Fatalf("Unexpected length of block devices list")
+	}
+}

